packages/plugin-core/src/features/windowDecorations.ts

```diff
@@ -126,114 +121,108 @@ export function delayedUpdateDecorations(
+    // Warn for missing or bad frontmatter and broken wikilinks
+    const allWarnings: Diagnostic[] = [];
+    const activeDecorations = new DefaultMap<
+      TextEditorDecorationType,
+      DecorationOptions[]
+    >(() => []);
+    // Only show decorations & warnings for notes
+    let note: NoteProps | undefined;
+    try {
+      note = VSCodeUtils.getNoteFromDocument(activeEditor.document);
+      if (_.isUndefined(note)) return {};
+    } catch (error) {
+      Logger.info({
+        ctx,
+        msg: "Unable to check if decorations should be updated",
+        error,
+      });
+      return {};
+    }
+    // Only decorate visible ranges, of which there could be multiple if the document is open in multiple tabs
+    const ranges = VSCodeUtils.mergeOverlappingRanges(
+      activeEditor.visibleRanges
+    );
+
+    for (const range of ranges) {
+      const decorateRange = VSCodeUtils.padRange({
+        range,
+        padding: VISIBLE_RANGE_MARGIN,
+        zeroCharacter: true,
+      });
+      const text = activeEditor.document.getText(decorateRange);
+      const proc = MDUtilsV5.procRemarkParse(
+        {
+          mode: ProcMode.FULL,
+          parseOnly: true,
+        },
+        {
+          dest: DendronASTDest.MD_DENDRON,
+          engine: getDWorkspace().engine,
+          vault: note.vault,
+          fname: note.fname,
+        }
+      );
+      const tree = proc.parse(text);
+      let frontmatter: FrontmatterContent | undefined;
+
+      visit(tree, (node) => {
+        const decorator = DECORATOR.get(node.type);
+        // Need to update node position with the added offset from the range
+        node.position!.start.line += decorateRange.start.line;
+        node.position!.end.line += decorateRange.start.line;
+        if (decorator) {
+          const decorations = decorator(node, activeEditor.document);
+          for (const { type, decoration } of decorations) {
+            activeDecorations.get(type).push(decoration);
+          }
+        }
+        if (node.type === DendronASTTypes.FRONTMATTER)
+          frontmatter = node as FrontmatterContent;
+      });
 
-export function updateDecorations(activeEditor: TextEditor) {
-  const ctx = "updateDecorations";
-  const text = activeEditor.document.getText();
-  // Warn for missing or bad frontmatter and broken wikilinks
-  const allWarnings: Diagnostic[] = [];
-  // Only show decorations & warnings for notes
-  let note: NoteProps | undefined;
-  try {
-    note = VSCodeUtils.getNoteFromDocument(activeEditor.document);
-    if (_.isUndefined(note)) return {};
-  } catch (error) {
+      if (decorateRange.start.line === 0) {
+        // Can't check frontmatter if frontmatter is not visible
+        if (_.isUndefined(frontmatter)) {
+          allWarnings.push(warnMissingFrontmatter(activeEditor.document));
+        } else {
+          allWarnings.push(
+            ...warnBadFrontmatterContents(activeEditor.document, frontmatter)
```

Comment:
can we rename this method? `warnBadFrontmatterContents` makes me think that is bad frontmatter. `checkAndWarnIfBadFrontmatterContents` would be more descriptive